src/browser/api_protocol/transport_strategy/elipc_strategy.ts

```diff
@@ -35,15 +35,21 @@ export class ElipcStrategy extends ApiTransportBase<MessagePackage> {
             if (strategyName !== this.constructor.name) {
                 next();
             } else if (typeof (action) === 'function') {
-                try {
-                    // singleFrameOnly check first so to prevent frame superceding when disabled.
-                    if (!data.singleFrameOnly === false || e.sender.isValidWithFrameConnect(e.frameRoutingId)) {
-                        action(identity, data, ack, nack);
-                    } else {
-                        nack('API access has been superseded by another frame in this window.');
-                    }
-                } catch (err) {
-                    nack(err);
+                // singleFrameOnly check first so to prevent frame superceding when disabled.
+                if (!data.singleFrameOnly === false || e.sender.isValidWithFrameConnect(e.frameRoutingId)) {
+                    Promise.resolve()
+                        .then(() => action(identity, data, ack, nack))
+                        .then(result => {
+                            if (result !== undefined) {
+                                ack(new AckPayload(result));
+                            }
+                        }).catch(err => {
+                            if (err !== undefined) {
```

Comment:
As before. Not Ack'ing or Nack'ing when the functions do not return a value is preserving the status-quo. Previously it has been up to the implementors of each API call to ensure that one or the other has been called. The new paradigm of returning a value (or promise) lessens the likelihood of a dangling API call.